torch/csrc/jit/passes/linear_to_matmul_add.cpp

```diff
@@ -0,0 +1,56 @@
+#include <algorithm>
+
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/passes/linear_to_matmul_add.h>
+#include <torch/csrc/jit/passes/tensorexpr_fuser.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+void DecomposeLinearToMatmulAdd(Block *b) {
+  // TODO only do the expansion for GPU since CPU fusion isn't on by default
+  // (and thus may slow down instead)
+  std::vector<Node*> linear_nodes;
+  for (auto it = b->nodes().begin(); it != b->nodes().end(); it++) {
+    for (auto* child_block : it->blocks()) {
+      DecomposeLinearToMatmulAdd(child_block);
+    }
+    if (it->kind() == aten::linear) {
+      linear_nodes.push_back(*it);
+    }
+  }
+
+  for (Node* node : linear_nodes) {
+    // If all its uses are supported for fusion, decompose into matmul + add
+    auto uses = node->output()->uses();
+    if (std::all_of(uses.begin(), uses.end(), [](const Use& u) {
+          return tensorexpr::isSupported(u.user);
+        })) {
+      auto g = b->owningGraph();
+      auto matmul_n =
+          g->create(
+               aten::matmul,
+               {node->namedInput("input"), node->namedInput("weight")},
+               1)
+              ->insertBefore(node);
+      auto add_n =
+          g->create(
+               aten::add, {matmul_n->output(), node->namedInput("bias")}, 1)
+              ->insertBefore(node);
+      // TODO aten::linear says bias is optional: what about that case?
```

Comment:
We should actually handle this case, and especially before actually enabling the pass. All we have to do is not append an `add` node to the `matmul` output, and then use the `matmul_n` Value * instead of `add_n` for replacing the node output.

Reply:
My understanding is the fusion that later can occur involves the add node. If no add node is present, do we still want to rewrite the linear node as a transpose + matmul?

Reply:
Sorry, yea that's correct if bias is not present there's no need to rewrite the `linear` node if bias is not present. I just mean `actually handle this case` as in, opt-out of rewriting the linear node if the bias isn't present.

Reply:
sounds good, doesn't have to be complete in first version and getting gpu environment set up is a pain (just dont to enable it initially in that case...)